File: source/sdks/sd-jwt-sdk/src/main/java/org/omnione/did/oid4vc/dcql/core/DCQLCredentialMatcher.java
/*
 * Copyright 2025 OmniOne.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.omnione.did.oid4vc.dcql.core;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import lombok.extern.slf4j.Slf4j;
import org.omnione.did.oid4vc.dcql.datamodel.DCQLQuery;
import org.omnione.did.oid4vc.exception.OID4VCException;
import org.omnione.did.sdjwt.datamodel.Disclosure;
import org.omnione.did.sdjwt.datamodel.SDJWT;
import org.omnione.did.sdjwt.util.SimpleJWTDecoder;

@Slf4j
public class DCQLCredentialMatcher {

  public static boolean matchesFormat(String requiredFormat) {
    if (requiredFormat == null) {
      return true;
    }

    Set<String> supportedFormats = Set.of("dc+sd-jwt", "vc+sd-jwt");

    boolean isSupported = supportedFormats.contains(requiredFormat);

    if (!isSupported) {
      log.info("Unsupported format: {}", requiredFormat);
    }

    return isSupported;
  }

  public static boolean matchesMetadata(SDJWT sdjwt, Map<String, Object> metadata) {
    if (metadata == null || metadata.isEmpty()) {
      return true;
    }

    try {
      if (metadata.containsKey("vct_values")) {
        List<String> requiredVcts = (List<String>) metadata.get("vct_values");
        if (!checkVctValues(sdjwt, requiredVcts)) {
          log.info("VCT value matching failed");
          return false;
        }
      }

      log.info("All metadata conditions satisfied");
      return true;

    } catch (OID4VCException | IllegalArgumentException e) {
      log.error("Error during metadata check: {}", e.getMessage(), e);
      return false;
    }
  }

  public static Set<String> extractMatchingClaimNames(DCQLQuery dcqlQuery, SDJWT sdjwt) {
    if (dcqlQuery == null || dcqlQuery.getCredentials() == null || sdjwt == null) {
      return Collections.emptySet();
    }

    Set<String> matchingClaims = new HashSet<>();

    Map<String, Object> allClaims = extractAllCredentialClaims(sdjwt);

    dcqlQuery.getCredentials().forEach(credential -> {
      if (credential.getClaims() == null) {
        matchingClaims.addAll(allClaims.keySet());
        log.info("credential.claims is null, including all claims: {}", allClaims.keySet());
        return;
      }

      if (credential.getClaims().isEmpty()) {
        log.info("credential.claims is empty, excluding claims");
        return;
      }

      credential.getClaims().forEach(claimQuery -> {
        List<Object> path = claimQuery.getPath();

        if (path == null || path.isEmpty()) {
          log.info("Claim path is empty or null");
          return;
        }

        processPathAndCollectClaims(allClaims, path, claimQuery, matchingClaims, sdjwt);
      });
    });

    return matchingClaims;
  }

  private static Map<String, Object> extractAllCredentialClaims(SDJWT sdjwt) {
    Map<String, Object> allClaims = new HashMap<>();

    try {
      SimpleJWTDecoder.SimpleJWT jwt = SimpleJWTDecoder.parse(sdjwt.getCredentialJwt());
      Map<String, Object> payload = jwt.getPayloadAsMap();

      payload.entrySet().stream().filter(entry -> !isReservedJWTClaim(entry.getKey()))
          .forEach(entry -> allClaims.put(entry.getKey(), entry.getValue()));

      log.info("Extracted claims from JWT payload: {}", allClaims.keySet());

    } catch (OID4VCException | IllegalArgumentException e) {
      log.error("JWT payload extraction failed: {}", e.getMessage(), e);
    }

    if (sdjwt.getDisclosures() != null && !sdjwt.getDisclosures().isEmpty()) {
      for (Disclosure disclosure : sdjwt.getDisclosures()) {
        if (disclosure.getClaimName() != null) {
          allClaims.put(disclosure.getClaimName(), disclosure.getClaimValue());
          log.info("Added top-level disclosed claim: {} = {}", disclosure.getClaimName(),
              disclosure.getClaimValue());
        }
      }

      integrateDisclosuresIntoCredential(allClaims, sdjwt);
    }

    log.info("All extracted claims after integrating disclosures: {}", allClaims.keySet());
    return allClaims;
  }

  private static void integrateDisclosuresIntoCredential(Map<String, Object> allClaims,
      SDJWT sdjwt) {
    Map<String, Disclosure> digestToDisclosure = new HashMap<>();
    for (Disclosure disclosure : sdjwt.getDisclosures()) {
      String digest = disclosure.digest();
      digestToDisclosure.put(digest, disclosure);
      log.info("Disclosure digest: {} -> {}.{}", digest, disclosure.getClaimName(),
          disclosure.getClaimValue());
    }

    for (Map.Entry<String, Object> entry : new HashMap<>(allClaims).entrySet()) {
      String claimName = entry.getKey();
      Object claimValue = entry.getValue();

      if (claimValue instanceof Map) {
        integrateDisclosuresIntoObject(claimName, (Map<String, Object>) claimValue,
            digestToDisclosure, allClaims);
      }
    }
  }

  private static void integrateDisclosuresIntoObject(String parentName,
      Map<String, Object> parentObject, Map<String, Disclosure> digestToDisclosure,
      Map<String, Object> allClaims) {

    Object sdValue = parentObject.get("_sd");
    if (sdValue instanceof List) {
      List<?> sdArray = (List<?>) sdValue;

      for (Object sdItem : sdArray) {
        if (sdItem instanceof String) {
          String digest = (String) sdItem;

          Disclosure disclosure = digestToDisclosure.get(digest);
          if (disclosure != null && disclosure.getClaimName() != null) {
            String claimName = disclosure.getClaimName();
            Object claimValue = disclosure.getClaimValue();

            parentObject.put(claimName, claimValue);
            log.info("Integrated disclosure: {}.{} = {}", parentName, claimName, claimValue);

            if (claimValue instanceof Map) {
              integrateDisclosuresIntoObject(parentName + "." + claimName,
                  (Map<String, Object>) claimValue, digestToDisclosure, allClaims);
            }
          }
        }
      }
    }
  }

  private static void processPathAndCollectClaims(Map<String, Object> allClaims, List<Object> path,
      DCQLQuery.ClaimQuery claimQuery, Set<String> matchingClaims, SDJWT sdjwt) {

    if (path.isEmpty() || !(path.get(0) instanceof String)) {
      log.info("Invalid path: first element must be a string (top-level claim)");
      return;
    }

    String topLevelClaim = (String) path.get(0);
    Object rootValue = allClaims.get(topLevelClaim);

    log.info("Processing path: {}, topLevelClaim: {}, rootValue type: {}", path, topLevelClaim,
        rootValue != null ? rootValue.getClass().getSimpleName() : "null");

    if (rootValue == null) {
      log.info("Top-level claim not found: {}", topLevelClaim);
      return;
    }

    if (path.size() == 1) {
      if (meetsClaimConditions(claimQuery, rootValue)) {
        matchingClaims.add(topLevelClaim);
        log.info("Single-level path matched: {} = {}", topLevelClaim, rootValue);
      } else {
        log.info("Single-level path condition not met: {} = {}", topLevelClaim, rootValue);
      }
      return;
    }

    List<Object> remainingPath = new ArrayList<>(path.subList(1, path.size()));
    log.info("Multi-depth path detected: {} with remaining path: {}", topLevelClaim, remainingPath);
    collectMatchingValuesFromPath(rootValue, remainingPath, claimQuery, topLevelClaim,
        matchingClaims);
  }

  private static void collectMatchingValuesFromPath(Object current, List<Object> remainingPath,
      DCQLQuery.ClaimQuery claimQuery, String claimNamePrefix, Set<String> matchingClaims) {

    log.info("collectMatchingValuesFromPath: claimNamePrefix={}, remainingPath={}, currentType={}",
        claimNamePrefix, remainingPath,
        current != null ? current.getClass().getSimpleName() : "null");

    if (remainingPath.isEmpty()) {
      log.info("End of path reached for: {}, value: {}", claimNamePrefix, current);
      if (meetsClaimConditions(claimQuery, current)) {
        matchingClaims.add(claimNamePrefix);
        log.info("Deep path matched: {} = {}", claimNamePrefix, current);
      }
      return;
    }

    Object nextPathElement = remainingPath.get(0);
    List<Object> nextRemaining = new ArrayList<>(remainingPath.subList(1, remainingPath.size()));

    if (nextPathElement == null) {
      if (!(current instanceof List)) {
        log.info("Wildcard path element (null) but current is not a list: {}", current.getClass());
        return;
      }

      List<?> currentList = (List<?>) current;
      for (int i = 0; i < currentList.size(); i++) {
        Object item = currentList.get(i);
        String newClaimName = claimNamePrefix + "[" + i + "]";
        collectMatchingValuesFromPath(item, nextRemaining, claimQuery, newClaimName,
            matchingClaims);
      }
      return;
    }

    if (nextPathElement instanceof Integer) {
      if (!(current instanceof List)) {
        log.info("Integer index but current is not a list: {}", current.getClass());
        return;
      }

      int idx = (Integer) nextPathElement;
      List<?> currentList = (List<?>) current;

      if (idx < 0 || idx >= currentList.size()) {
        log.info("Index out of bounds: {} (list size: {})", idx, currentList.size());
        return;
      }

      Object element = currentList.get(idx);
      String newClaimName = claimNamePrefix + "[" + idx + "]";
      collectMatchingValuesFromPath(element, nextRemaining, claimQuery, newClaimName,
          matchingClaims);
      return;
    }

    if (nextPathElement instanceof String) {
      if (!(current instanceof Map)) {
        log.info("String key but current is not a map: type={}, value={}",
            current.getClass().getSimpleName(), current);
        return;
      }

      Map<?, ?> currentMap = (Map<?, ?>) current;
      String key = (String) nextPathElement;

      log.info("Searching for key '{}' in map with keys: {}", key, currentMap.keySet());

      Object nextValue = currentMap.get(key);
      if (nextValue == null) {
        log.info("Key not found in map: {} (available keys: {})", key, currentMap.keySet());
        return;
      }

      String newClaimName = claimNamePrefix + "." + key;
      log.info("Found nested value at path '{}': {}", newClaimName, nextValue);
      collectMatchingValuesFromPath(nextValue, nextRemaining, claimQuery, newClaimName,
          matchingClaims);
    }
  }

  private static boolean checkVctValues(SDJWT sdjwt, List<String> requiredVcts) {
    if (requiredVcts == null || requiredVcts.isEmpty()) {
      return true;
    }

    try {
      SimpleJWTDecoder.SimpleJWT jwt = SimpleJWTDecoder.parse(sdjwt.getCredentialJwt());
      Map<String, Object> payload = jwt.getPayloadAsMap();

      Object vctClaim = payload.get("vct");
      if (vctClaim == null) {
        log.info("SD-JWT has no vct claim");
        return false;
      }

      String actualVct = vctClaim.toString();
      boolean matches = requiredVcts.contains(actualVct);

      log.info("VCT matching: required={}, actual={}, result={}", requiredVcts, actualVct, matches);

      return matches;

    } catch (OID4VCException | IllegalArgumentException e) {
      log.error("VCT check error: {}", e.getMessage(), e);
      return false;
    }
  }

  private static boolean meetsClaimConditions(DCQLQuery.ClaimQuery claimQuery, Object actualValue) {
    if (actualValue == null) {
      return false;
    }

    if (claimQuery.getValues() != null && !claimQuery.getValues().isEmpty()) {
      boolean valueMatches = claimQuery.getValues().contains(actualValue);
      log.info("  values condition: {} in {} = {}", actualValue, claimQuery.getValues(),
          valueMatches);
      if (!valueMatches) {
        return false;
      }
    }

    if (claimQuery.getValue() != null) {
      boolean exactMatch = claimQuery.getValue().equals(actualValue);
      log.info("  value condition: {} == {} = {}", actualValue, claimQuery.getValue(), exactMatch);
      if (!exactMatch) {
        return false;
      }
    }

    if (claimQuery.getMin() != null) {
      if (!checkMinCondition(actualValue, claimQuery.getMin())) {
        return false;
      }
    }

    if (claimQuery.getMax() != null) {
      if (!checkMaxCondition(actualValue, claimQuery.getMax())) {
        return false;
      }
    }

    return true;
  }

  private static boolean checkMinCondition(Object actualValue, Object minValue) {
    try {
      if (actualValue instanceof Number && minValue instanceof Number) {
        double actual = ((Number) actualValue).doubleValue();
        double min = ((Number) minValue).doubleValue();
        return actual >= min;
      }

      if (actualValue instanceof String && minValue instanceof String) {
        return ((String) actualValue).compareTo((String) minValue) >= 0;
      }

      return true;

    } catch (OID4VCException | IllegalArgumentException e) {
      return false;
    }
  }

  private static boolean checkMaxCondition(Object actualValue, Object maxValue) {
    try {
      if (actualValue instanceof Number && maxValue instanceof Number) {
        double actual = ((Number) actualValue).doubleValue();
        double max = ((Number) maxValue).doubleValue();
        return actual <= max;
      }

      if (actualValue instanceof String && maxValue instanceof String) {
        return ((String) actualValue).compareTo((String) maxValue) <= 0;
      }

      return true;

    } catch (OID4VCException | IllegalArgumentException e) {
      return false;
    }
  }

  private static boolean isReservedJWTClaim(String claimName) {
    Set<String> reservedClaims = Set.of("iss", "sub", "aud", "exp", "nbf", "iat", "jti", "_sd_alg",
        "_sd", "cnf", "vct");
    return reservedClaims.contains(claimName);
  }
}

File: source/apps/ios-app/oid4vc/oid4vc/Data/DTO/DCQL/DCQLCredentialMatcher.swift
//
/*
 * Copyright 2026 OmniOne.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Foundation
import CryptoKit

// MARK: - AnyJSON (lossless JSON value container)

public enum AnyJSON: Codable, Equatable, Hashable {
    case null
    case bool(Bool)
    case number(Double)          // JSON numbers; keep as Double
    case string(String)
    case array([AnyJSON])
    case object([String: AnyJSON])

    // Convenience accessors
    public var asBool: Bool? { if case .bool(let v) = self { return v } else { return nil } }
    public var asDouble: Double? { if case .number(let v) = self { return v } else { return nil } }
    public var asString: String? { if case .string(let v) = self { return v } else { return nil } }
    public var asArray: [AnyJSON]? { if case .array(let v) = self { return v } else { return nil } }
    public var asObject: [String: AnyJSON]? { if case .object(let v) = self { return v } else { return nil } }

    // Convert to Foundation types for JSONSerialization / path navigation
    public func toFoundation() -> Any {
        switch self {
        case .null: return NSNull()
        case .bool(let b): return b
        case .number(let n): return n
        case .string(let s): return s
        case .array(let a): return a.map { $0.toFoundation() }
        case .object(let o): return o.mapValues { $0.toFoundation() }
        }
    }

    // Create from Foundation JSON types
    public static func fromFoundation(_ value: Any) -> AnyJSON? {
        if value is NSNull { return .null }
        if let b = value as? Bool { return .bool(b) }
        if let n = value as? NSNumber {
            // NSNumber can represent Bool too; make sure we didn't already catch Bool above
            return .number(n.doubleValue)
        }
        if let s = value as? String { return .string(s) }
        if let a = value as? [Any] {
            return .array(a.compactMap { AnyJSON.fromFoundation($0) })
        }
        if let o = value as? [String: Any] {
            var dict: [String: AnyJSON] = [:]
            for (k, v) in o {
                if let j = AnyJSON.fromFoundation(v) {
                    dict[k] = j
                } else {
                    // Non-JSON value -> drop as null (shouldn't happen with JSONSerialization)
                    dict[k] = .null
                }
            }
            return .object(dict)
        }
        return nil
    }

    // MARK: Codable

    public init(from decoder: Decoder) throws {
        let c = try decoder.singleValueContainer()

        if c.decodeNil() {
            self = .null
            return
        }

        if let b = try? c.decode(Bool.self) {
            self = .bool(b)
            return
        }

        // Decode numbers as Double. This preserves JSON numeric semantics.
        if let n = try? c.decode(Double.self) {
            self = .number(n)
            return
        }

        if let s = try? c.decode(String.self) {
            self = .string(s)
            return
        }

        if let a = try? c.decode([AnyJSON].self) {
            self = .array(a)
            return
        }

        if let o = try? c.decode([String: AnyJSON].self) {
            self = .object(o)
            return
        }

        throw DecodingError.typeMismatch(
            AnyJSON.self,
            .init(codingPath: decoder.codingPath, debugDescription: "Unsupported JSON value")
        )
    }

    public func encode(to encoder: Encoder) throws {
        var c = encoder.singleValueContainer()

        switch self {
        case .null:
            try c.encodeNil()
        case .bool(let b):
            try c.encode(b)
        case .number(let n):
            try c.encode(n)
        case .string(let s):
            try c.encode(s)
        case .array(let a):
            try c.encode(a)
        case .object(let o):
            try c.encode(o)
        }
    }

    // MARK: Hashable (canonical JSON)

    public func hash(into hasher: inout Hasher) {
        // Stable canonical JSON ensures hash stability for object key ordering etc.
        hasher.combine(canonicalJSONString())
    }

    private func canonicalJSONString() -> String {
        // Encode as canonical JSON (sorted keys for objects).
        // Not the fastest, but robust and deterministic for hashing/equality validation use-cases.
        switch self {
        case .null:
            return "null"
        case .bool(let b):
            return b ? "true" : "false"
        case .number(let n):
            // Use JSON-style minimal representation where possible.
            // Avoid scientific notation differences by using a normalized string.
            if n.isNaN || n.isInfinite { return "null" } // JSON can't represent these
            // Strip trailing .0 if integer-like
            if floor(n) == n { return String(format: "%.0f", n) }
            // Use a reasonable precision
            return String(n)
        case .string(let s):
            return AnyJSON.escapeJSONString(s)
        case .array(let a):
            return "[" + a.map { $0.canonicalJSONString() }.joined(separator: ",") + "]"
        case .object(let o):
            let keys = o.keys.sorted()
            let parts = keys.map { key -> String in
                let k = AnyJSON.escapeJSONString(key)
                let v = o[key]?.canonicalJSONString() ?? "null"
                return "\(k):\(v)"
            }
            return "{" + parts.joined(separator: ",") + "}"
        }
    }

    private static func escapeJSONString(_ s: String) -> String {
        // Minimal JSON string escaping
        var out = "\""
        for ch in s.unicodeScalars {
            switch ch.value {
            case 0x22: out += "\\\""        // "
            case 0x5C: out += "\\\\"        // \
            case 0x08: out += "\\b"
            case 0x0C: out += "\\f"
            case 0x0A: out += "\\n"
            case 0x0D: out += "\\r"
            case 0x09: out += "\\t"
            default:
                if ch.value < 0x20 {
                    out += String(format: "\\u%04X", ch.value)
                } else {
                    out += String(ch)
                }
            }
        }
        out += "\""
        return out
    }
}

// MARK: - DCQL Path Element

public enum DCQLPathElement: Codable, Hashable {
    case key(String)
    case index(Int)
    case wildcard   // corresponds to null in Java path list

    public init(from decoder: Decoder) throws {
        let c = try decoder.singleValueContainer()
        if c.decodeNil() { self = .wildcard; return }
        if let i = try? c.decode(Int.self) { self = .index(i); return }
        if let s = try? c.decode(String.self) { self = .key(s); return }
        throw DecodingError.typeMismatch(
            DCQLPathElement.self,
            .init(codingPath: decoder.codingPath, debugDescription: "Path element must be string, int, or null")
        )
    }

    public func encode(to encoder: Encoder) throws {
        var c = encoder.singleValueContainer()
        switch self {
        case .wildcard: try c.encodeNil()
        case .index(let i): try c.encode(i)
        case .key(let s): try c.encode(s)
        }
    }
}



public enum DCQLCredentialMatcher {

    public static func matchesFormat(requiredFormat: String?) -> Bool {
        guard let requiredFormat = requiredFormat else { return true }
        let supported: Set<String> = ["dc+sd-jwt", "vc+sd-jwt"]
        return supported.contains(requiredFormat)
    }

    public static func matchesMetadata(sdjwt: SDJWT, metadata: [String: AnyJSON]?) -> Bool {
        guard let metadata = metadata, !metadata.isEmpty else { return true }

        if let v = metadata["vct_values"] {
            // vct_values must be array of strings
            guard let required = v.asArray?.compactMap({ $0.asString }) else { return false }
            return checkVctValues(sdjwt: sdjwt, requiredVcts: required)
        }

        return true
    }

    private static func checkVctValues(sdjwt: SDJWT, requiredVcts: [String]?) -> Bool {
        guard let requiredVcts = requiredVcts, !requiredVcts.isEmpty else { return true }
        do {
            let jwt = try SimpleJWTDecoder.parse(sdjwt.credentialJwt)
            guard let vct = jwt.payload["vct"] else { 
                print("❌ 'vct' claim missing in JWT payload")
                return false 
            }
            let actual = String(describing: vct)
            print("--- checkVctValues ---")
            print("Required VCTs: \(requiredVcts)")
            print("Actual VCT: \(actual)")
            let isContained = requiredVcts.contains(actual)
            print("Match result: \(isContained)")
            print("-----------------------")
            return isContained
        } catch {
            print("❌ SimpleJWTDecoder failed: \(error)")
            return false
        }
    }

    public static func extractMatchingClaimNames(dcqlQuery: DCQLQuery?, sdjwt: SDJWT?) -> Set<String> {
        guard let dcqlQuery = dcqlQuery, let creds = dcqlQuery.credentials, let sdjwt = sdjwt else {
            return []
        }

        var matching: Set<String> = []
        let allClaims = extractAllCredentialClaims(sdjwt: sdjwt)

        for credential in creds {
            if credential.claims == nil {
                matching.formUnion(allClaims.keys)
                continue
            }

            for claimQuery in credential.claims ?? [] {
                guard let path = claimQuery.path, !path.isEmpty else { continue }
                processPathAndCollectClaims(allClaims: allClaims,
                                           path: path,
                                           claimQuery: claimQuery,
                                           matchingClaims: &matching)
            }
        }

        return matching
    }

    private static func extractAllCredentialClaims(sdjwt: SDJWT) -> [String: Any] {
        var allClaims: [String: Any] = [:]

        if let jwt = try? SimpleJWTDecoder.parse(sdjwt.credentialJwt) {
            for (k, v) in jwt.payload where !isReservedJWTClaim(k) {
                allClaims[k] = v
            }
        }

        if !sdjwt.disclosures.isEmpty {
            for d in sdjwt.disclosures {
                if let name = d.claimName {
                    allClaims[name] = d.claimValue as Any
                }
            }
            integrateDisclosuresIntoCredential(allClaims: &allClaims, sdjwt: sdjwt)
        }

        return allClaims
    }

    private static func integrateDisclosuresIntoCredential(allClaims: inout [String: Any], sdjwt: SDJWT) {
        var digestToDisclosure: [String: Disclosure] = [:]
        for d in sdjwt.disclosures {
            digestToDisclosure[d.digest()] = d
        }

        let snapshot = allClaims
        for (claimName, claimValue) in snapshot {
            if var obj = claimValue as? [String: Any] {
                integrateDisclosuresIntoObject(parentName: claimName,
                                              parentObject: &obj,
                                              digestToDisclosure: digestToDisclosure)
                allClaims[claimName] = obj
            }
        }
    }

    private static func integrateDisclosuresIntoObject(parentName: String,
                                                       parentObject: inout [String: Any],
                                                       digestToDisclosure: [String: Disclosure]) {
        guard let sd = parentObject["_sd"] as? [Any] else { return }

        for item in sd {
            guard let digest = item as? String else { continue }
            guard let disclosure = digestToDisclosure[digest], let claimName = disclosure.claimName else { continue }

            let claimValue = disclosure.claimValue as Any
            parentObject[claimName] = claimValue

            if var nested = claimValue as? [String: Any] {
                integrateDisclosuresIntoObject(parentName: parentName + "." + claimName,
                                              parentObject: &nested,
                                              digestToDisclosure: digestToDisclosure)
                parentObject[claimName] = nested
            }
        }
    }

    private static func processPathAndCollectClaims(allClaims: [String: Any],
                                                    path: [DCQLPathElement],
                                                    claimQuery: DCQLQuery.ClaimQuery,
                                                    matchingClaims: inout Set<String>) {

        guard case .key(let topLevel) = path[0] else { return }
        guard let rootValue = allClaims[topLevel] else { return }

        if path.count == 1 {
            if meetsClaimConditions(claimQuery: claimQuery, actualValue: rootValue) {
                matchingClaims.insert(topLevel)
            }
            return
        }

        let remaining = Array(path.dropFirst())
        collectMatchingValuesFromPath(current: rootValue,
                                      remainingPath: remaining,
                                      claimQuery: claimQuery,
                                      claimNamePrefix: topLevel,
                                      matchingClaims: &matchingClaims)
    }

    private static func collectMatchingValuesFromPath(current: Any,
                                                      remainingPath: [DCQLPathElement],
                                                      claimQuery: DCQLQuery.ClaimQuery,
                                                      claimNamePrefix: String,
                                                      matchingClaims: inout Set<String>) {
        if remainingPath.isEmpty {
            if meetsClaimConditions(claimQuery: claimQuery, actualValue: current) {
                matchingClaims.insert(claimNamePrefix)
            }
            return
        }

        let next = remainingPath[0]
        let nextRemaining = Array(remainingPath.dropFirst())

        switch next {
        case .wildcard:
            guard let list = current as? [Any] else { return }
            for i in 0..<list.count {
                let newName = "\(claimNamePrefix)[\(i)]"
                collectMatchingValuesFromPath(current: list[i],
                                              remainingPath: nextRemaining,
                                              claimQuery: claimQuery,
                                              claimNamePrefix: newName,
                                              matchingClaims: &matchingClaims)
            }

        case .index(let idx):
            guard let list = current as? [Any] else { return }
            guard idx >= 0, idx < list.count else { return }
            let newName = "\(claimNamePrefix)[\(idx)]"
            collectMatchingValuesFromPath(current: list[idx],
                                          remainingPath: nextRemaining,
                                          claimQuery: claimQuery,
                                          claimNamePrefix: newName,
                                          matchingClaims: &matchingClaims)

        case .key(let key):
            guard let dict = current as? [String: Any] else { return }
            guard let nextValue = dict[key] else { return }
            let newName = "\(claimNamePrefix).\(key)"
            collectMatchingValuesFromPath(current: nextValue,
                                          remainingPath: nextRemaining,
                                          claimQuery: claimQuery,
                                          claimNamePrefix: newName,
                                          matchingClaims: &matchingClaims)
        }
    }

    private static func meetsClaimConditions(claimQuery: DCQLQuery.ClaimQuery, actualValue: Any) -> Bool {
        // Convert actualValue to AnyJSON for robust comparisons
        let actualJSON = AnyJSON.fromFoundation(actualValue) ?? .null

        if let values = claimQuery.values, !values.isEmpty {
            if !values.contains(actualJSON) { return false }
        }

        if let v = claimQuery.value {
            if actualJSON != v { return false }
        }

        if let min = claimQuery.min, !checkMinCondition(actualValue: actualJSON, minValue: min) {
            return false
        }
        if let max = claimQuery.max, !checkMaxCondition(actualValue: actualJSON, maxValue: max) {
            return false
        }

        return true
    }

    private static func checkMinCondition(actualValue: AnyJSON, minValue: AnyJSON) -> Bool {
        switch (actualValue, minValue) {
        case (.number(let a), .number(let m)):
            return a >= m
        case (.string(let a), .string(let m)):
            return a.compare(m) != .orderedAscending
        default:
            return true
        }
    }

    private static func checkMaxCondition(actualValue: AnyJSON, maxValue: AnyJSON) -> Bool {
        switch (actualValue, maxValue) {
        case (.number(let a), .number(let m)):
            return a <= m
        case (.string(let a), .string(let m)):
            return a.compare(m) != .orderedDescending
        default:
            return true
        }
    }

    private static func isReservedJWTClaim(_ claimName: String) -> Bool {
        let reserved: Set<String> = ["iss","sub","aud","exp","nbf","iat","jti","_sd_alg","_sd","cnf","vct"]
        return reserved.contains(claimName)
    }
}

